### ISSUE TITLE
Lack of Diagnostic Information when Bug present in Upstream Provider

### ISSUE THREAD

**Issue by zbuchheit** [pulumi/pulumi-terraform-bridge#2413]
### What happened?

When an upstream provider has a bug that triggers an `inconsistent result of apply` error, our downstream providers displays this behavior as a spurious diff without any error or diagnostics. I would expect the behavior to display a log or show what is causing the spurious diff.

### Example Error
```
2024-09-11T17:09:09.291-0700 [ERROR] vertex "mist_device_gateway.gateway_one" error: Provider produced inconsistent result after apply
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to mist_device_gateway.gateway_one, provider
│ "provider[\"registry.terraform.io/juniper/mist\"]" produced an unexpected new value:
│ .additional_config_cmds: was cty.ListValEmpty(cty.String), but now null.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue
│ tracker.
```

When I translate the code to pulumi in a bridged provider instead of an error I get a spurious diff and doesn't appear in the logs even when TF_LOG=debug is used.

### Output of `pulumi about`

N/A

### Additional context

I think the behavior itself is actually acceptable in that it doesn't throw an error and is an upstream bug that needs to be fixed, but I would expect a diagnostic log to show up to help troubleshoot this or some sort of warning.

Currently it is very difficult to figure out what is causing the spurious diff if the resource has a complex schema, whereas in TF the error points you to the problem immediately

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by guineveresaenger** [comment]
Thank you for filing this bug. I've added kind/enhancement as well since this is a subcategory of inconsistencies in diff display that we may have a better fix for soon.

cc @iwahbe

**Comment by zbuchheit** [comment]
If I can be of any help to provide logs etc, just let me know. Thanks! 😄

**Comment by rshade** [comment]
I just ran into this with another client, is it possible to raise up the error to the client instead of the diff. If the tf provider is generating an error we should capture and raise that up a level.

**Comment by VenelinMartinov** [comment]
@rshade in the original issue Zach reported it's the TF CLI raising the error, not the provider.

Can you please include some more detail on what you hit with the client? It doesn't sound like the same issue if it's the provider raising the error.

**Comment by VenelinMartinov** [comment]
Note the error in the issue is not produced by the TF provider but by the TF CLI.

We might consider adding a similar check in the bridge which verifies that the state returned by Create/Update is consistent with the plan before that. We'd need to make sure this is backwards-compatible though, so it could be opt-in or debug-only.

**Comment by rshade** [comment]
They are getting a diff each up, whether things change or not, so I don't think backward compatibility will be a problem.

**Comment by VenelinMartinov** [comment]
The bridge is used for all of our bridged providers, so any changes there would need to work for all of them. The suggestion for making this opt-in so to maintain backwards-compatibility in other providers. Some of them might be currently depending on this behaviour.

**Comment by t0yv0** [comment]
> We might consider adding a similar check in the bridge which verifies that the state returned by Create/Update is consistent with the plan before that. We'd need to make sure this is backwards-compatible though, so it could be opt-in or debug-only.

I've just commented on the PR @rshade has put up, this is possibly a bad idea since the checks are already implemented in the TF codebase and re-implementing it here signs us up for maintaining a shadow copy that may get out of sync with the original.

Rather, the question is at which point we interface with this functionality incorrectly, possibly around the LegacyTypeSystem setting.

**Comment by t0yv0** [comment]
The code is probably somewhere around here: https://github.com/opentofu/opentofu/blob/4083fe2cc4a483fc19d9769e0eb19ebee2386b33/internal/tofu/node_resource_abstract_instance.go#L1048

If we could find a way to vendor more of that code and activate it so we continue to reuse upstream code maximally that could be interesting.
